cluster-controller/src/main/java/io/strimzi/controller/cluster/ClusterController.java

```diff
@@ -52,30 +54,24 @@
     private final KubernetesClient client;
     private final Map<String, String> labels;
     private final String namespace;
-    private ConfigMapOperations configMapOperations;
-    private StatefulSetOperations statefulSetOperations;
-    private DeploymentOperations deploymentOperations;
-    private DeploymentConfigOperations deploymentConfigOperations;
+    private final ConfigMapOperations configMapOperations;
+    private final StatefulSetOperations statefulSetOperations;
+    private final DeploymentOperations deploymentOperations;
+    private final DeploymentConfigOperations deploymentConfigOperations;
 
     private Watch configMapWatch;
 
     private long reconcileTimer;
-    private KafkaClusterOperations kafkaClusterOperations;
-    private KafkaConnectClusterOperations kafkaConnectClusterOperations;
-    private KafkaConnectS2IClusterOperations kafkaConnectS2IClusterOperations;
+    private final KafkaClusterOperations kafkaClusterOperations;
+    private final KafkaConnectClusterOperations kafkaConnectClusterOperations;
+    private final KafkaConnectS2IClusterOperations kafkaConnectS2IClusterOperations;
 
-    public ClusterController(ClusterControllerConfig config) {
+    public ClusterController(Vertx vertx, ClusterControllerConfig config) {
```

Comment:
I'm a little bit confused. What's the need for having the `Vertx` instance as parameter (It's passed by the main using the Vert.x instance which starts the verticle). A `vertx` field is already available (as protected) in the base `AbstractVerticle` class and should be the right one but ... as @tombentley said me ... "I'm teaching grandmother to suck eggs" so there is something I'm missing ;)

Reply:
@ppatierno the way we use a ClusterController it's only ever associated with a single vertx instance. The `protected` field in `AbstractVerticle` is null at construction time, which meant we had to make all the different `ResourceOperations` classes non-`final`, and their null-ness depended on both whether the verticle had been started and whether we were running against OpenShift.

By passing in the vertx instance the verticle will be started in we can move all those resources into the constructor and make the fields final. This means we get the benefit of `javac`'s definite assignment analysis, and it's simpler to reason about the fields null-ness.

Reply:
Oh, another benefit is it allows us to test the ClusterController without actually deploying it as a verticle. Why is that useful? Well it means that we don't have to factor out the watcher+reconciliation stuff into a different class in order to test the `ClusterController`. If we could only access the vertx when deployed as a verticle then @scholzj's point about that refactoring would be motivated by testing.

Reply:
@tombentley now I remember that it was a bug I fixed moving the operations in the `start` method of the verticle due to having the `Vertx` instance null in the constructor. I can't understand the benefit of having `ResourceOperations` as `final` but I can understand the second one related to testing even if it seems that developing in a "test-driven" way doesn't bring to a really good design in this case imho. At least, we should document why the `Vert.x` instance is passed there otherwise more other people could think that such a parameter is useless and raising the same point.

Reply:
> I can't understand the benefit of having ResourceOperations as `final`

Generally speaking, I prefer for fields to be `final` if possible, since it helps eliminate all kinds of potential bugs, such as where someone accidentally reassigns a field which is supposed to be only set once. FWIW, further refactoring has resulted in `Main` getting the job of wiring together all the Operations, and ClusterController ctor now just gets passed the ClusterOperations.